## Target: 
* Help developers understand what type of apps are likely to attract more users on Google Play and the App Store.
* object: apps that are free to download and install, and our main source of revenue consists of in-app ads. This means that the number of users of our apps determines our revenue for any given app.

## Dataset:
* A dataset containing data about approximately 10,000 Android apps from Google Play; the data was collected in August 2018. 
* A dataset containing data about approximately 7,000 iOS apps from the App Store; the data was collected in July 2017.
* Source [link](https://www.kaggle.com/lava18/google-play-store-apps), [link](https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps)

In [1]:
def open_file(dataset):
    file= open(dataset,encoding='utf8')
    from csv import reader
    read_file=reader(file)
    return list(read_file)

ios_header=open_file('C:/Users/Admin/Duyen/AppleStore.csv')
ios=ios_header[1:]
androi_header=open_file('C:/Users/Admin/Duyen/googleplaystore.csv')
androi=androi_header[1:]


### opening and exploring these two data sets 

In [2]:
def explore_data(dataset, start, end, row_columns=True):
    dataset_slice=dataset[start:end]
    for row in dataset_slice: # instead of print dataset_slice, print row with line space help easier to look at
        print(row)
        print('\n') # adds a new (empty) line after each row

    
    if row_columns:
        print('number of row:', len(dataset))
        print('number of column:',len(dataset[0]))

print(explore_data(ios_header,0,2,True))
print(explore_data(androi_header,0,2,True))

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


number of row: 7198
number of column: 16
None
['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


number of row: 10842
number of column: 13
None


## Clean data
we only build apps that are free to download and install, and we design them for an English-speaking audience. This means that we'll need to do the following:

* Remove non-English apps like 爱奇艺PPS -《欢乐颂2》电视剧热播.

Find if any row's rate is greater than 5

In [3]:
def rate_greater_than_5(dataset,index_rate,index_name):
    list_name=[]   
    for row in dataset:
        rate=float(row[index_rate])
        name=row[index_name]
        if rate>5:
            list_name.append(name)
    return(list_name)

rate_greater_than_5(ios,7,1)
rate_greater_than_5(androi,2,0)
        
    

['Life Made WI-Fi Touchscreen Photo Frame']

In [4]:
print(androi[10472])

['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


The rating is 19. This is clearly off because the maximum rating for a Google Play app is 5=> delete row

In [5]:
print(len(androi))
del androi[10472]
print(len(androi))

10841
10840


## Find duplicate app

In [6]:
def duplicate(dataset):
    unique_app=[]
    duplicate_app=[]
    for row in dataset:
        name=row[0]
        if name in unique_app:
            duplicate_app.append(name)
        else:
            unique_app.append(name)
    return duplicate_app

def list_duplicate(column):
    list_dup={}
    for value in column:
            if value in list_dup:
                list_dup[value]+=1
            else:
                list_dup[value]=1
    return print(list_dup)

dup_ios=duplicate(ios)
print('number of duplicate ios:',len(dup_ios))
dup_androi=duplicate(androi)
print('number of duplicate ios:',len(dup_androi))
list_duplicate(dup_ios)
list_duplicate(dup_androi)

number of duplicate ios: 0
number of duplicate ios: 1181
{}
{'Quick PDF Scanner + OCR FREE': 2, 'Box': 2, 'Google My Business': 2, 'ZOOM Cloud Meetings': 1, 'join.me - Simple Meetings': 2, 'Zenefits': 1, 'Google Ads': 2, 'Slack': 2, 'FreshBooks Classic': 1, 'Insightly CRM': 1, 'QuickBooks Accounting: Invoicing & Expenses': 2, 'HipChat - Chat Built for Teams': 1, 'Xero Accounting Software': 1, 'MailChimp - Email, Marketing Automation': 1, 'Crew - Free Messaging and Scheduling': 1, 'Asana: organize team projects': 1, 'Google Analytics': 1, 'AdWords Express': 1, 'Accounting App - Zoho Books': 1, 'Invoice & Time Tracking - Zoho': 1, 'Invoice 2go — Professional Invoices and Estimates': 1, 'SignEasy | Sign and Fill PDF and other Documents': 1, 'Genius Scan - PDF Scanner': 1, 'Tiny Scanner - PDF Scanner App': 1, 'Fast Scanner : Free PDF Scan': 1, 'Mobile Doc Scanner (MDScan) Lite': 1, 'TurboScan: scan documents and receipts in PDF': 1, 'Tiny Scanner Pro: PDF Doc Scan': 1, 'Docs To Go™ Free Of

above, frequency of apps has been deducteded 1 from the dataset. Below is example for Instagram app.

In [7]:
for row in androi:
    name=row[0]
    if name=='Instagram':
        print(row)

['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66509917', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


The main difference happens on the fourth position of each row, which corresponds to the number of reviews. The different numbers show the data was collected at different times.
We can use this information to build a criterion for removing the duplicates. The higher the number of reviews, the more recent the data should be. Rather than removing duplicates randomly, we'll only keep the row with the highest number of reviews and remove the other entries for any given app.

In [8]:
review_max={}
for row in androi:
    name=row[0]
    reviews=float(row[3])
    if name in review_max and review_max[name]<reviews:
        review_max[name]=reviews
    if name not in review_max:
        review_max[name]=reviews
print(len(review_max))
print(len(androi)-1181)                              

9659
9659


Now, let's use the reviews_max dictionary to remove the duplicates. For the duplicate cases, we'll only keep the entries with the highest number of reviews. In the code cell below:

In [9]:
clean_androi=[]
already_added=[]
for row in androi:
    name=row[0]
    n_reviews=float(row[3])
    if (review_max[name]==n_reviews) and (name not in already_added):
        clean_androi.append(row)
        already_added.append(name)

explore_data(clean_androi, 0, 2, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


number of row: 9659
number of column: 13


We have 9659 rows, just as expected.
## Removing Non-English Apps

All these characters that are specific to English texts are encoded using the ASCII standard. Each ASCII character has a corresponding number between 0 and 127 associated with it, and we can take advantage of that to build a function that checks an app name and tells us whether it contains non-ASCII characters.

We built this function below, and we use the built-in `ord()` function to find out the corresponding encoding number of each character.

realize that the reduce of apps is quite at big amount for iso from 7,198 to 5,707, about 20% reduction.The function seems to work fine, but some English app names use emojis or other symbols (™, — (em dash), – (en dash), etc.) that fall outside of the ASCII range. Because of this, we'll remove useful apps if we use the function in its current form. To minimize the impact of data loss, we'll only remove an app if its name has more than three non-ASCII characters:

In [10]:
def is_english(string):
    non_ascii = 0
    
    for character in string:
        if ord(character) > 127:
            non_ascii += 1
    
    if non_ascii > 3:
        return False
    else:
        return True

androi_english=[]
ios_english=[]

for row in clean_androi:
    name=row[0]
    if is_english(name):
        androi_english.append(row)


for row in ios:
    name=row[1]
    if is_english(name):
        ios_english.append(row)
        
explore_data(androi_english, 0, 1, True)
print('\n')
explore_data(ios_english, 0, 1, True)
            

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


number of row: 9614
number of column: 13


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


number of row: 6183
number of column: 16


## Free app

Found that price is 'string' and can't covert totally to float ( $4.95 can't convert to float).

In [11]:
androi_final = []
ios_final = []

for app in androi_english:
    price = app[7]
    if price == '0':
        androi_final.append(app)
        
for app in ios_english:
    price = app[4]
    if price == '0.0':
        ios_final.append(app)
        
print(len(androi_final))
print(len(ios_final))
            


8864
3222


## Most Common Apps by Genre

In [12]:
def freq_table(dataset, index):
    table = {}
    
    for row in dataset:
        value = row[index]
        if value in table:
            table[value] += 1
        else:
            table[value] = 1
    
    table_percentages = {}
    for key in table:
        percentage = (table[key] / len(dataset)) * 100
        table_percentages[key] = percentage 
    
    return table_percentages


def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)
        
    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])

In [13]:
display_table(ios_final, -5)

Games : 58.16263190564867
Entertainment : 7.883302296710118
Photo & Video : 4.9658597144630665
Education : 3.662321539416512
Social Networking : 3.2898820608317814
Shopping : 2.60707635009311
Utilities : 2.5139664804469275
Sports : 2.1415270018621975
Music : 2.0484171322160147
Health & Fitness : 2.0173805090006205
Productivity : 1.7380509000620732
Lifestyle : 1.5828677839851024
News : 1.3345747982619491
Travel : 1.2414649286157666
Finance : 1.1173184357541899
Weather : 0.8690254500310366
Food & Drink : 0.8069522036002483
Reference : 0.5586592178770949
Business : 0.5276225946617008
Book : 0.4345127250155183
Navigation : 0.186219739292365
Medical : 0.186219739292365
Catalogs : 0.12414649286157665


For IOS, games apps percentage is highest, at 58%, following by entertainment, photo & video. But it doesn't present it has highest number of users.

In [14]:
display_table(androi_final, -4)

Tools : 8.449909747292418
Entertainment : 6.069494584837545
Education : 5.347472924187725
Business : 4.591606498194946
Productivity : 3.892148014440433
Lifestyle : 3.892148014440433
Finance : 3.7003610108303246
Medical : 3.531137184115524
Sports : 3.463447653429603
Personalization : 3.3167870036101084
Communication : 3.2378158844765346
Action : 3.1024368231046933
Health & Fitness : 3.0798736462093865
Photography : 2.944494584837545
News & Magazines : 2.7978339350180503
Social : 2.6624548736462095
Travel & Local : 2.3240072202166067
Shopping : 2.2450361010830324
Books & Reference : 2.1435018050541514
Simulation : 2.0419675090252705
Dating : 1.861462093862816
Arcade : 1.8501805054151623
Video Players & Editors : 1.7712093862815883
Casual : 1.7599277978339352
Maps & Navigation : 1.3989169675090252
Food & Drink : 1.2409747292418771
Puzzle : 1.128158844765343
Racing : 0.9927797833935018
Role Playing : 0.9363718411552346
Libraries & Demo : 0.9363718411552346
Auto & Vehicles : 0.9250902527075

For androi, tools and entertainment is at highest rate at about 8.5% and 6% correspondingly. but there are more catergories in androi comparing to ios. for example, regarding to games,there are 8 different catergories games.

## Most Popular Apps by Genre
### IOS

In [15]:
def genre_view(dataset, review_index, genre_index):
    genre_ios=freq_table(dataset, genre_index)
    for key in genre_ios:
        total=0
        for row in dataset:
            n_review=float(row[review_index])
            genre=row[genre_index]
            if genre==key:
                total+=1
                genre_ios[key]+=n_review
        avg_n_ratings=genre_ios[key]/total
        print(key, ':', avg_n_ratings)

genre_view(ios_final,5,-5)

Social Networking : 71548.38009322698
Photo & Video : 28441.57478662322
Games : 22788.70072712482
Music : 57326.56133965352
Reference : 74942.14214773432
Health & Fitness : 23298.0464212386
Weather : 52279.92389376607
Utilities : 18684.487826746674
Travel : 28243.831036623218
Shopping : 26919.721512813692
News : 21248.05429243717
Navigation : 86090.36436995656
Lifestyle : 16485.79574250557
Entertainment : 14029.861745284632
Food & Drink : 33333.95411354629
Sports : 23008.92958734785
Book : 39758.531036623215
Finance : 31467.97548106766
Education : 7004.014087470672
Productivity : 21028.44175090893
Business : 7491.148683682039
Catalogs : 4004.031036623215
Medical : 612.0310366232153


Genre: navigation, reference and social networking have the highest viewers in average

In [16]:
for app in ios_final:
    if app[-5] == 'Navigation':
        print(app[1], ':', app[5]) # print name and number of ratings

Waze - GPS Navigation, Maps & Real-time Traffic : 345046
Google Maps - Navigation & Transit : 154911
Geocaching® : 12811
CoPilot GPS – Car Navigation & Offline Maps : 3582
ImmobilienScout24: Real Estate Search in Germany : 187
Railway Route Search : 5


On average, navigation apps have the highest number of user reviews, but this figure is heavily influenced by Waze and Google Maps, which have close to half a million user reviews together

In [17]:
for app in ios_final:
    if app[-5] == 'Reference':
        print(app[1], ':', app[5])

Bible : 985920
Dictionary.com Dictionary & Thesaurus : 200047
Dictionary.com Dictionary & Thesaurus for iPad : 54175
Google Translate : 26786
Muslim Pro: Ramadan 2017 Prayer Times, Azan, Quran : 18418
New Furniture Mods - Pocket Wiki & Game Tools for Minecraft PC Edition : 17588
Merriam-Webster Dictionary : 16849
Night Sky : 12122
City Maps for Minecraft PE - The Best Maps for Minecraft Pocket Edition (MCPE) : 8535
LUCKY BLOCK MOD ™ for Minecraft PC Edition - The Best Pocket Wiki & Mods Installer Tools : 4693
GUNS MODS for Minecraft PC Edition - Mods Tools : 1497
Guides for Pokémon GO - Pokemon GO News and Cheats : 826
WWDC : 762
Horror Maps for Minecraft PE - Download The Scariest Maps for Minecraft Pocket Edition (MCPE) Free : 718
VPN Express : 14
Real Bike Traffic Rider Virtual Reality Glasses : 8
教えて!goo : 0
Jishokun-Japanese English Dictionary & Translator : 0


### Android

In [18]:
genre_view(androi_final,3,-4)

Art & Design : 26119.12448913562
Art & Design;Creativity : 5677.844614921781
Auto & Vehicles : 14140.291769393325
Beauty : 7476.237696682787
Books & Reference : 87995.07970264107
Business : 24239.738554315718
Comics : 43364.696466773625
Comics;Creativity : 516.0112815884477
Communication : 995608.4746962225
Dating : 21953.284008861174
Education : 16211.395247833174
Education;Creativity : 6384.261281588448
Education;Education : 234564.87794825513
Education;Pretend Play : 25151.411281588447
Education;Brain Games : 192591.01128158843
Entertainment : 103389.25849348436
Entertainment;Brain Games : 79104.29699587416
Entertainment;Creativity : 143559.34461492178
Entertainment;Music & Video : 79679.54461492179
Events : 2555.852551429717
Finance : 38535.910671832346
Food & Drink : 57478.80219067935
Health & Fitness : 78094.98197755913
House & Home : 26435.477035013104
Libraries & Demo : 10925.81851050411
Lifestyle : 33611.47794825512
Lifestyle;Pretend Play : 140995.01128158843
Card : 166334.986

communication has the highest the viewer

In [19]:
for app in androi_final:
    if app[-4] == 'Communication':
        print(app[0], ':', app[3])

WhatsApp Messenger : 69119316
Messenger for SMS : 125257
My Tele2 : 158679
imo beta free calls and text : 659395
Contacts : 66602
Call Free – Free Call : 30209
Web Browser & Explorer : 36901
Browser 4G : 192948
MegaFon Dashboard : 99559
ZenUI Dialer & Contacts : 437674
Cricket Visual Voicemail : 13698
TracFone My Account : 20769
Xperia Link™ : 45487
TouchPal Keyboard - Fun Emoji & Android Keyboard : 615381
Skype Lite - Free Video Call & Chat : 33053
My magenta : 42370
Android Messages : 781810
Google Duo - High Quality Video Calls : 2083237
Seznam.cz : 46702
Antillean Gold Telegram (original version) : 2939
AT&T Visual Voicemail : 13761
GMX Mail : 258556
Omlet Chat : 40751
My Vodacom SA : 25021
Microsoft Edge : 27187
Messenger – Text and Video Chat for Free : 56646578
imo free video calls and chat : 4785988
Calls & Text by Mo+ : 83239
free video calls and chat : 594728
Skype - free IM & video calls : 10484169
Who : 2451093
GO SMS Pro - Messenger, Free Themes, Emoji : 2876500
Messaging+

## Conclusion:
* for IOS app, navigation, reference and social networking have the highest viewers in average.
* for Andoid, commnucation has the higest viewers
* there is not clear genre catagories in Androi, to really find the most popular genre and it frequency.
* While social networking and commnucation are in highest viewers rates but their app quantity is low. The compnay can take time to invest deeper in this area